app/controllers/challenges_controller.rb

```diff
@@ -49,28 +51,28 @@ def create
   end
 
   def show
-    @challenge = Challenge.find(params['id']) rescue nil
-    return redirect_to root_path if @challenge.nil?
-
-    # TODO, there is a better way to do this...
-    users = @challenge.entries.map {|e| e.user_id }.uniq
-    @users = User.where(:_id.in => users).inject({}) {|h,u| h[u.id] = u; h}
+    # Limit to id to avoid downloading uneccessary entries
+    challenge_id = params['id']
+    challenge = Challenge.only(:id).find(challenge_id) rescue nil
+    return redirect_to root_path if challenge.nil?
 
     respond_to do |format|
-      format.json {
-        render :json => {
-          'in' => {
-            'data' => @challenge.input,
-            'type' => @challenge.input_type
-          },
-          'out' => {
-            'data' => @challenge.output,
-            'type' => @challenge.output_type
-          },
-          'client' => Vimgolf::VERSION
-      }}
-
-      format.html
+      format.json { render :json => json_show(challenge_id) }
+
+      format.html {
+        # TODO, there is a better way to do this...
```

Comment:
Re, TODO: what do you have in mind here?

Reply:
It is your TODO :) just moved in previous PR

But I keep it, because we can totally use RepositoryChallenge.paginate_leaderboard to load the page in one query and less ruby